tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AspNetCore/EndpointsCollection/KestrelServerImplStartAsyncIntegration.cs

```diff
@@ -0,0 +1,104 @@
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+
+#if !NETFRAMEWORK
+
+using System;
+using System.ComponentModel;
+using System.Threading;
+using Datadog.Trace.AppSec;
+using Datadog.Trace.ClrProfiler.CallTarget;
+using Datadog.Trace.Configuration;
+using Datadog.Trace.DuckTyping;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.AspNetCore.EndpointsCollection;
+
+/// <summary>
+/// System.Threading.Tasks.Task Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerImpl::StartAsync[TContext](Microsoft.AspNetCore.Hosting.Server.IHttpApplication`1[TContext],System.Threading.CancellationToken) calltarget instrumentation
+/// </summary>
+[InstrumentMethod(
+    AssemblyName = "Microsoft.AspNetCore.Server.Kestrel.Core",
+    TypeName = "Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerImpl",
+    MethodName = "StartAsync",
+    ReturnTypeName = ClrNames.Task,
+    ParameterTypeNames = ["Microsoft.AspNetCore.Hosting.Server.IHttpApplication`1[!!0]", ClrNames.CancellationToken],
+    MinimumVersion = "5.0.0",
+    MaximumVersion = SupportedVersions.LatestDotNet,
+    InstrumentationCategory = InstrumentationCategory.AppSec,
+    IntegrationName = nameof(IntegrationId.AspNetCore))]
+[InstrumentMethod(
+    AssemblyName = "Microsoft.AspNetCore.Server.Kestrel.Core",
+    TypeName = "Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServer",
+    MethodName = "StartAsync",
+    ReturnTypeName = ClrNames.Task,
+    ParameterTypeNames = ["Microsoft.AspNetCore.Hosting.Server.IHttpApplication`1[!!0]", ClrNames.CancellationToken],
+    MinimumVersion = "2.2.0",
+    MaximumVersion = "3.*.*",
+    InstrumentationCategory = InstrumentationCategory.AppSec,
+    IntegrationName = nameof(IntegrationId.AspNetCore))]
+[Browsable(false)]
+[EditorBrowsable(EditorBrowsableState.Never)]
+public class KestrelServerImplStartAsyncIntegration
+{
+    private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor<KestrelServerImplStartAsyncIntegration>();
+
+    internal static CallTargetState OnMethodBegin<TTarget, TApplication>(TTarget instance, ref TApplication? application, ref CancellationToken cancellationToken)
+    {
+        try
+        {
+            if (Security.Instance.ApiSecurity.CanCollectEndpoints())
+            {
+                GatherEndpoints(instance);
+            }
+        }
+        catch (Exception ex)
+        {
+            Log.Error(ex, "API Security: Failed to collect endpoints.");
+        }
+
+        return CallTargetState.GetDefault();
+    }
+
+    private static void GatherEndpoints(object? instance)
+    {
+        if (instance == null)
+        {
+            Log.Warning("API Security: Endpoints collection: No Instance.");
+            return;
+        }
+
+        if (!instance.TryDuckCast<IKestrelServer>(out var kestrelServer))
```

Comment:
I guess you could cast `instance` to `IKestrelServer` in a where clause in `OnMethodBegin`